data/transactions/logic/debugger.go

```diff
@@ -0,0 +1,272 @@
+	}
+	ds.Globals = globals
+
+	// pre-allocate state maps
+	if (cx.runModeFlags & runModeApplication) != 0 {
+		ds.GlobalStateChanges = make(basics.StateDelta)
+
+		// allocate maximum possible slots in the hashmap even if Txn.Accounts might have duplicate entries
+		locals := 1 + len(cx.Txn.Txn.Accounts) // sender + referenced accounts
+		ds.LocalStateChanges = make(map[basics.Address]basics.StateDelta, locals)
+
+		// do not pre-allocate ds.LocalStateChanges[addr] since it initialized during update
+	}
+
+	return ds
+}
+
+// LineToPC converts line to pc
+// Return 0 on unsuccess
+func (d *DebugState) LineToPC(line int) int {
+	if len(d.PCOffset) == 0 || line < 1 {
+		return 0
+	}
+
+	lines := strings.Split(d.Disassembly, "\n")
+	if line > len(lines) {
+		return 0
+	}
+	offset := len(strings.Join(lines[:line], "\n"))
+
+	for i := 0; i < len(d.PCOffset); i++ {
+		if d.PCOffset[i].Offset >= offset {
+			return d.PCOffset[i].PC
+		}
+	}
+	return 0
+}
+
+// PCToLine converts pc to line
+// Return 0 on unsuccess
+func (d *DebugState) PCToLine(pc int) int {
+	if len(d.PCOffset) == 0 {
+		return 0
+	}
+
+	offset := 0
+	for i := 0; i < len(d.PCOffset); i++ {
+		if d.PCOffset[i].PC >= pc {
+			offset = d.PCOffset[i].Offset
+			break
+		}
+	}
+
+	one := 1
+	// handle end of the program
+	if offset == 0 {
+		offset = d.PCOffset[len(d.PCOffset)-1].Offset
+		one = 0
+	}
+	if offset > len(d.Disassembly) {
+		return 0
+	}
+
+	return len(strings.Split(d.Disassembly[:offset], "\n")) - one
+}
+
+func stackValueToTealValue(sv *stackValue) basics.TealValue {
+	tv := sv.toTealValue()
+	return basics.TealValue{
+		Type:  tv.Type,
+		Bytes: base64.StdEncoding.EncodeToString([]byte(tv.Bytes)),
+		Uint:  tv.Uint,
+	}
+}
+
+func (cx *evalContext) refreshDebugState() *DebugState {
+	ds := &cx.debugState
+
+	// Update pc, line, error, stack, and scratch space
+	ds.PC = cx.pc
+	ds.Line = ds.PCToLine(cx.pc)
+	if cx.err != nil {
+		ds.Error = cx.err.Error()
+	}
+
+	stack := make([]basics.TealValue, len(cx.stack), len(cx.stack))
+	for i, sv := range cx.stack {
+		stack[i] = stackValueToTealValue(&sv)
+	}
+
+	scratch := make([]basics.TealValue, len(cx.scratch), len(cx.scratch))
+	for i, sv := range cx.scratch {
+		scratch[i] = stackValueToTealValue(&sv)
+	}
+
+	ds.Stack = stack
+	ds.Scratch = scratch
+
+	if (cx.runModeFlags & runModeApplication) != 0 {
+		if cx.globalStateCow != nil {
+			for k, v := range cx.globalStateCow.delta {
+				ds.GlobalStateChanges[k] = v
+			}
+		}
+		for addr, cow := range cx.localStateCows {
+			delta := make(basics.StateDelta)
+			for k, v := range cow.cow.delta {
+				delta[k] = v
+			}
+			ds.LocalStateChanges[addr] = delta
+		}
+	}
+
+	return ds
+}
+
+func (dbg *WebDebuggerHook) postState(state *DebugState, endpoint string) error {
+	var body bytes.Buffer
+	enc := protocol.NewJSONEncoder(&body)
+	err := enc.Encode(state)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf("%s/%s", dbg.URL, endpoint), &body)
```

Comment:
1. "MethodPost" -> http.MethodPost
2. Use the url.URL to build the url here instead of formatting it yourself using Sprintf

Reply:
done, thanks